### PR TITLE
[Feature] Receipt trait in alloy-consensus

### LINKED ISSUE #466: [Feature] Receipt trait in alloy-network
Followup work: 
- `trait Receipt` in alloy-network
- Receipt bound on `type Recept` in `trait Network`
- `fn status` in `trait Receipt`
- `impl Receipt for ReceiptEnvelope` and related types
- change this function to a delegate to `Receipt`

_Originally posted by @prestwich in https://github.com/alloy-rs/alloy/pull/456#discussion_r1552786585_

### PR DESCRIPTION
**Pull request by developeruche**
## Motivation
 Closes #466

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

**Comment by developeruche** [comment]
@prestwich *change status function on rpc receipt type to delegate to Receipt implementation of inner* 

This is not quite possible at the moment;

pub enum ReceiptEnvelope<T = Log>  ----> This uses Log from `use alloy_primitives::{Bloom, Log};`

and 

pub struct TransactionReceipt<T = ReceiptEnvelope<Log>> ----> This uses Log from `rpc-type` making the trait bound not satified... that's is `Receipt implementation of inner` is not accessiable

File: crates/consensus/src/receipt/any.rs [hunk 1]
```diff
@@ -39,31 +39,28 @@ impl AnyReceiptEnvelope {
...
         }
     }
+}
 
-    /// Return true if the transaction was successful.
-    pub const fn is_success(&self) -> bool {
-        self.status()
-    }
-
+impl TxReceipt for AnyReceiptEnvelope {
```
Comment (prestwich) [inline crates/consensus/src/receipt/any.rs:44]: this should now be generic over `T` 

`impl<T> TxReceipt for AnyReceiptEnvelope<T>`

File: crates/consensus/src/receipt/any.rs [hunk 2]
```diff
@@ -39,31 +39,28 @@ impl AnyReceiptEnvelope {
...
+
     /// Returns the cumulative gas used at this receipt.
-    pub const fn cumulative_gas_used(&self) -> u64 {
+    fn cumulative_gas_used(&self) -> u64 {
         self.inner.receipt.cumulative_gas_used
     }
 
     /// Return the receipt logs.
-    pub fn logs(&self) -> &[Log] {
+    fn logs(&self) -> &[Log] {
```
Comment (prestwich) [inline crates/consensus/src/receipt/any.rs:61]: this will probably need to be `-> &[T]`

File: crates/consensus/src/receipt/envelope.rs [hunk 3]
```diff
@@ -93,6 +68,31 @@ impl<T> ReceiptEnvelope<T> {
     }
 }
 
+impl TxReceipt for ReceiptEnvelope {
```
Comment (prestwich) [inline crates/consensus/src/receipt/envelope.rs:71]: same, this should be generic over `T`

File: crates/network/src/lib.rs [hunk 4]
```diff
@@ -58,7 +59,7 @@ pub trait Network: Clone + Copy + Sized + Send + Sync + 'static {
     type UnsignedTx;
 
     /// The network receipt envelope type.
-    type ReceiptEnvelope: Eip2718Envelope;
+    type ReceiptEnvelope: Eip2718Envelope + TxReceipt;
```
Comment (prestwich) [inline crates/network/src/lib.rs:62]: the `type ReceiptResponse` below should also have a `+ TxReceipt` bound

**Review by prestwich — CHANGES_REQUESTED** [review]

**Review by prestwich — APPROVED** [review]

File: crates/consensus/src/receipt/receipts.rs [hunk 5]
```diff
@@ -78,7 +78,7 @@ pub struct ReceiptWithBloom<T = Log> {
 }
 
 impl TxReceipt for ReceiptWithBloom {
-    fn success(&self) -> bool {
```
Comment (prestwich) [inline crates/consensus/src/receipt/receipts.rs:81]: followup work:
- Make this impl generic over <T>

**Pull request merged**